llvm/lib/Target/X86/X86InstrCompiler.td

```diff
@@ -689,27 +689,106 @@ def : Pat<(v8f64 (X86cmov VR512:$t, VR512:$f, timm:$cond, EFLAGS)),
+  // FP32-specific CTSELECT for i386 (FP->GR32->FP conversion approach)
+  multiclass CTSELECT_I386_FP32<RegisterClass RC> {
+    let hasNoSchedulingInfo = 1, FPForm = SpecialFP in {
+      def rr : PseudoI<(outs RC:$dst, GR8:$tmp_byte, GR32:$tmp_mask, GR32:$tmp_a, GR32:$tmp_b),
+                       (ins RC:$src1, RC:$src2, i8imm:$cond), []> {
+        let Constraints =
+            "@earlyclobber $dst,@earlyclobber $tmp_byte,@earlyclobber "
+            "$tmp_mask,@earlyclobber $tmp_a,@earlyclobber $tmp_b";
+      }
+    }
+  }
+  // FP64-specific CTSELECT for i386 (FP->2xGR32->FP conversion approach)
+  multiclass CTSELECT_I386_FP64<RegisterClass RC> {
+    let hasNoSchedulingInfo = 1, FPForm = SpecialFP in {
+      def rr : PseudoI<(outs RC:$dst, GR8:$tmp_byte, GR32:$tmp_mask, 
+                             GR32:$tmp_a_lo, GR32:$tmp_a_hi, GR32:$tmp_b_lo, GR32:$tmp_b_hi),
+                       (ins RC:$src1, RC:$src2, i8imm:$cond), []> {
+        let Constraints =
+            "@earlyclobber $dst,@earlyclobber $tmp_byte,@earlyclobber "
+            "$tmp_mask,@earlyclobber $tmp_a_lo,@earlyclobber $tmp_a_hi,"
+            "@earlyclobber $tmp_b_lo,@earlyclobber $tmp_b_hi";
+      }
+    }
+  }
+  // FP80-specific CTSELECT for i386 (sequential processing approach to reduce register pressure)
+  multiclass CTSELECT_I386_FP80<RegisterClass RC> {
+    let hasNoSchedulingInfo = 1, FPForm = SpecialFP in {
+      def rr : PseudoI<(outs RC:$dst, GR8:$tmp_byte, GR32:$tmp_mask, GR32:$tmp_work),
+                       (ins RC:$src1, RC:$src2, i8imm:$cond), []> {
+        let Constraints =
+            "@earlyclobber $dst,@earlyclobber $tmp_byte,@earlyclobber "
+            "$tmp_mask,@earlyclobber $tmp_work";
+      }
+    }
+  }
+
+  // Legacy CTSELECT for other targets (unchanged)
   multiclass CTSELECT_NOCMOV<RegisterClass RC, ValueType VT> {
     let hasNoSchedulingInfo = 1 in {
-      def rr  : PseudoI<(outs RC:$dst),
-                        (ins RC:$src1, RC:$src2, i8imm:$cond),
-                        [(set RC:$dst, ( VT (X86ctselect RC:$src1, RC:$src2, timm:$cond, EFLAGS)))]>;
-
-      // TODO: Do we need register-memory variant??
-      //def rm : PseudoI<(outs t.RegClass:$dst),
-      //                 (ins t.RegClass:$src1, t.MemOperand:$src2, i8imm:$cond),
-      //                 [(set t.RegClass:$dst, (X86ctselect t.RegClass:$src1, (t.LoadNode addr:$src2), timm:$cond, EFLAGS))]>;
+      def rr : PseudoI<(outs RC:$dst), (ins RC:$src1, RC:$src2, i8imm:$cond),
+                       [(set RC:$dst, (VT(X86ctselect RC:$src1, RC:$src2,
+                                          timm:$cond, EFLAGS)))]>;
     }
   }
 }
 
+// Post-RA expansion pseudos for i386 (no custom inserter)
+let isCodeGenOnly = 1, hasSideEffects = 1, ForceDisassemble = 1 in {
+  let Predicates = [NoCMOV] in {
+    defm CTSELECT_I386_GR16 : CTSELECT_I386<GR16, GR8>;
+    defm CTSELECT_I386_GR32 : CTSELECT_I386<GR32, GR8>;
+  }
+  let Predicates = [NoCMOV, HasMMX] in {
```

Comment:
Do we need `NoCMOV` predicate here? Why not support it if arch has CMOV as well? Most modern x86-64 variant use sse operations but there are a few older variant that may have cmov and use mmx for vector types.